### PR TITLE
Implement Seq2seq

### PR DESCRIPTION
**Pull request by dding3**
## What changes were proposed in this pull request?
1. Implement a seq2seq layer
2. support set/getGradHiddenState/s in Recurrent/RecurrentDecoder
3. add example to use seq2seq

## How was this patch tested?
ut

### CODE REVIEW

**Comment by dding3** [comment]
There are some duplicate code with pr https://github.com/intel-analytics/BigDL/pull/1591 . With rebase this  PR after https://github.com/intel-analytics/BigDL/pull/1591, https://github.com/intel-analytics/BigDL/pull/1619 are merged.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/TextToLabeledSentence.scala [hunk 1]
```diff
@@ -60,6 +60,30 @@ class TextToLabeledSentence[T: ClassTag](dictionary: Dictionary)
...
+class TextToLabeledSentence2[T: ClassTag](dictionary: Dictionary)
+                                        (implicit ev: TensorNumeric[T])
+  extends Transformer[Array[String], LabeledSentence[T]] {
+  private val buffer = new LabeledSentence[T]()
+
+  override def apply(prev: Iterator[Array[String]]): Iterator[LabeledSentence[T]] = {
+    prev.map(sentence => {
+      val indexes = sentence.map(x =>
+        ev.fromType[Int](dictionary.getIndex(x)))
+      val nWords = indexes.length / 2.0
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/TextToLabeledSentence.scala]: If nWords is an odd integer, will drop the middle word in sentence ?
Is it correct?

File: docs/docs/APIGuide/Layers/Recurrent-Layers.md [hunk 2]
```diff
@@ -1901,20 +1904,663 @@ println(output)
...
+`forward(input)` is expected to be a 5D tensor (`batch x time x outputLen x outputPlane x length x height x width`).
+
+Ref.
+
+1. http://arxiv.org/abs/1409.3215
+
+Parameters:
+
+* `encoderRecs` List of Recurrents used for encodering
+* `decoderRecs` List of Recurrents/RecurrentDecoders userd for decodering
```
Comment (qiuxin2012) [inline docs/docs/APIGuide/Layers/Recurrent-Layers.md]: userd ?

File: pyspark/bigdl/nn/layer.py [hunk 3]
```diff
@@ -5085,10 +5100,43 @@ class ConvLSTMPeephole3D(Layer):
...
         super(ConvLSTMPeephole3D, self).__init__(None, bigdl_type, input_size, output_size, kernel_i, kernel_c, stride,
-                                                 padding, wRegularizer, uRegularizer, bRegularizer, cRegularizer, with_peephole)
+                                                 padding, wRegularizer, uRegularizer, bRegularizer, cRegularizer, with_peephole,
+                                                 memory_optim)
 
+class Seq2seq(Layer):
+    '''
+|   Sequence to sequence mode.
+|   Ref.
+|   A.: https://arxiv.org/abs/1506.04214 (blueprint for this module)
```
Comment (qiuxin2012) [inline pyspark/bigdl/nn/layer.py]: Your Ref. in readme is `http://arxiv.org/abs/1409.3215`

**Pull request closed without merging**